### PR TITLE
refactor(permission-controller): Rationalize permission and caveat validation

### PR DESCRIPTION
**Pull request by rekmarks**
## Explanation

Via discussion during https://github.com/MetaMask/metamask-extension/pull/27847, it became apparent that there was confusion re: the purpose of permission vs. caveat validators and when to use one vs. the other. This has now been properly documented in multiple places in the permission controller package.

In addition, the permission controller implementation contained some irregularities regarding its own permission validation logic. This surfaced in the form of redundant permission and caveat validator calls, both of which have now been removed. (In addition, one overly complicated "optimization" to avoid redundant caveat validator calls has been removed.)

Reviewers: although I doubt it's a problem, we should take care to ensure that we weren't relying on the old behavior in any of our existing caveat or permission specifications.

## References

- https://github.com/MetaMask/metamask-extension/pull/27847#discussion_r1879060900

### `@metamask/permission-controller`

- **CHANGED**: Remove redundant caveat validator calls
  - In some cases, caveats were being validated multiple times or without the possibility of having changed.
  - The intended purpose of permission and caveat validators has also been documented. See `ARCHITECTURE.md`.

## Checklist

- [x] I've updated the test suite for new or updated code as appropriate
- [x] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [x] I've highlighted breaking changes using the "BREAKING" category above as appropriate
- [x] I've prepared draft pull requests for clients and consumer packages to resolve any breaking changes

### CODE REVIEW

File: packages/permission-controller/src/PermissionController.ts [hunk 1]
```diff
@@ -1773,17 +1797,10 @@ export class PermissionController<
...
-
       if (specification.factory) {
         permission = specification.factory(permissionOptions, requestData);
       } else {
         permission = constructPermission(permissionOptions);
-
-        // We do not need to validate caveats in this case, because the plain
-        // permission constructor function does not modify the caveats, which
-        // were already validated by `constructCaveats` above.
-        performCaveatValidation = false;
```
Comment (rekmarks) [inline packages/permission-controller/src/PermissionController.ts:1786]: We're all trying to find the guy who introduced this futile and brittle optimization.

Comment (FrederikBolding) [reply packages/permission-controller/src/PermissionController.ts:1786]: image

File: packages/permission-controller/src/PermissionController.ts [hunk 2]
```diff
@@ -1829,10 +1846,10 @@ export class PermissionController<
     specification: PermissionSpecificationConstraint,
     permission: PermissionConstraint,
     origin: OriginString,
-    { invokePermissionValidator, performCaveatValidation } = {
```
Comment (FrederikBolding) [inline packages/permission-controller/src/PermissionController.ts:1832]: Hmm, why do we wanna remove the default values here?

**Review by FrederikBolding — APPROVED** [review]
LGTM - have verified that Snaps doesn't use permission validators to validate caveat contents too.

**Review by Gudahtt — APPROVED** [review]
LGTM!

**Review by jiexi — APPROVED** [review]

**Pull request merged**